### PR TITLE
Added TYPE(native) data type for external tables

### PR DESCRIPTION
**Pull request by paul-rogers**
Druid recently added support for nested JSON types. The JSON-style of `extern()` can label input columns as a JSON object or array using the Druid `complex<json>` type. This PR adds an equivalent of the type to SQL using the `TYPE` keyword. Example:

```sql
INSERT INTO dst SELECT *
FROM TABLE(http(uris => ARRAY['http://foo.com/bar.json'],
                format => 'csv'))
     (x VARCHAR, y VARCHAR, z TYPE('COMPLEX<json>'))
PARTITIONED BY ALL TIME
```

The semantics are that the `z` column above is _some_ complex JSON type, but we don't care which: Druid's indexer will figure it out. Since we don't know the exact type (other than that it isn't simple), we let Druid do the work by declaring that the input column type varies: it is of Druid's type `COMPLEX<json>`. At run time, Druid will pick the actual type, which may be `COMPLEX<json>` or may be something else.

The argument to `TYPE()` is the name of a Druid type exactly as Druid expects it: upper case `COMPLEX` and lower case `json`.

#### Release note

See the comments above and the modified `reference.md` file.

<hr>

This PR has:

- [X] been self-reviewed.
- [X] added documentation for new or modified features or behaviors.
- [X] a release note entry in the PR description.
- [X] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [X] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [ ] been tested in a test Druid cluster.

### CODE REVIEW

**Review by vogievetsky — APPROVED** [review]
I approve the API part

File: sql/src/main/codegen/config.fmpp [hunk 1]
```diff
@@ -68,6 +69,7 @@ data: {
       "CLUSTERED"
       "OVERWRITE"
       "PARTITIONED"
+      "VARIANT"
```
Comment (clintropolis) [inline sql/src/main/codegen/config.fmpp:72]: this seems stale

File: sql/src/main/codegen/includes/common.ftl [hunk 2]
```diff
@@ -92,3 +92,18 @@ SqlNodeList ClusterItems() :
     return new SqlNodeList(list, s.addAll(list).pos());
   }
 }
+
+SqlTypeNameSpec VariantType() :
```
Comment (clintropolis) [inline sql/src/main/codegen/includes/common.ftl:96]: nit: maybe rename something like `DruidTypePassthrough`?

File: sql/src/main/java/org/apache/druid/sql/calcite/planner/DruidTypeSystem.java [hunk 3]
```diff
@@ -35,6 +35,8 @@ public class DruidTypeSystem implements RelDataTypeSystem
    */
   public static final int DEFAULT_TIMESTAMP_PRECISION = 3;
 
+  public static final String VARIANT_TYPE_NAME = "VARIANT";
```
Comment (clintropolis) [inline sql/src/main/java/org/apache/druid/sql/calcite/planner/DruidTypeSystem.java:38]: nit: stale

File: sql/src/main/java/org/apache/druid/sql/calcite/external/Externals.java [hunk 4]
```diff
@@ -242,7 +242,11 @@ private static String convertType(String name, SqlDataTypeSpec dataType)
     if (typeName == null || !typeName.isSimple()) {
       throw unsupportedType(name, dataType);
     }
-    SqlTypeName type = SqlTypeName.get(typeName.getSimple());
+    String simpleName = typeName.getSimple();
+    if (StringUtils.toLowerCase(simpleName).startsWith(("complex<"))) {
+      return simpleName;
+    }
+    SqlTypeName type = SqlTypeName.get(simpleName);
```
Comment (clintropolis) [inline sql/src/main/java/org/apache/druid/sql/calcite/external/Externals.java:249]: nit: i know this isn't new, but these variable names are kinda confusing which initially had me wondering why we were looking up the type name of the type name 😅 but see now that `typeName` is an `SqlIdentifier` and `type` is an `SqlTypeName`.

**Review by clintropolis — COMMENTED** [review]
:+1: this seems useful to have the ability to pass through native druid type strings until we design a proper solution to give complex types SQL type names of their own

File: sql/src/main/java/org/apache/druid/sql/calcite/external/Externals.java (hunk 4 above)
Comment (paul-rogers) [reply sql/src/main/java/org/apache/druid/sql/calcite/external/Externals.java:249]: Changed all the names. I hope the new names are a bit more clear.

**Pull request merged**